buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -999,5 +1000,153 @@ protected void deallocate() {
+     */
+    public static boolean isText(ByteBuf buf, Charset charset) {
+        return isText(buf, buf.readerIndex(), buf.readableBytes(), charset);
+    }
+
+    /**
+     * Returns {@code true} if the specified {@link ByteBuf} starting at {@code index} with {@code length} is valid
+     * text using the given {@link Charset}, otherwise return {@code false}.
+     *
+     * @param buf The given {@link ByteBuf}.
+     * @param index The start index of the specified buffer.
+     * @param length The length of the specified buffer.
+     * @param charset The specified {@link Charset}.
+     *
+     * @throws IndexOutOfBoundsException if {@code index} + {@code length} is greater than {@code buf.readableBytes}
+     */
+    public static boolean isText(ByteBuf buf, int index, int length, Charset charset) {
+        checkNotNull(buf, "buf");
+        checkNotNull(charset, "charset");
+        final int maxIndex = buf.readerIndex() + buf.readableBytes();
+        if (index < 0 || length < 0 || index > maxIndex - length) {
+            throw new IndexOutOfBoundsException("index: " + index + " length: " + length);
+        }
+        if (charset.equals(CharsetUtil.UTF_8)) {
+            return isUTF8(buf, index, length);
+        } else if (charset.equals(CharsetUtil.US_ASCII)) {
+            return isASCII(buf, index, length);
+        } else {
+            CharsetDecoder decoder = CharsetUtil.decoder(charset, CodingErrorAction.REPORT, CodingErrorAction.REPORT);
+            try {
+                decoder.decode(buf.nioBuffer(index, length));
+                return true;
+            } catch (CharacterCodingException e) {
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Returns {@code true} if the specified {@link ByteBuf} starting at {@code index} with {@code length} is valid
+     * ASCII text, otherwise return {@code false}.
+     *
+     * @param buf The given {@link ByteBuf}.
+     * @param index The start index of the specified buffer.
+     * @param length The length of the specified buffer.
+     */
+    private static boolean isASCII(ByteBuf buf, int index, int length) {
+        final int endIndex = index + length;
+        while (index < endIndex) {
+            byte b = buf.getByte(index++);
+            if (b < 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns {@code true} if the specified {@link ByteBuf} starting at {@code index} with {@code length} is valid
+     * UTF8 text, otherwise return {@code false}.
+     *
+     * @param buf The given {@link ByteBuf}.
+     * @param index The start index of the specified buffer.
+     * @param length The length of the specified buffer.
+     *
+     * @see
+     * <a href=http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/96393e490afd/src/share/classes/sun/nio/cs/UTF_8.java#l37
+     * >UTF_8.java</a>
+     */
+    private static boolean isUTF8(ByteBuf buf, int index, int length) {
```

Comment:
> @windie we can not pull in code that is based on openjdk as the license is not compatible with ASL2. If you need to borrow some code please get it from android or apache harmony.

Just some comments are copied from the header of UTF_8.java for readability. Let me take a look at other projects and replace them.